packages/@aws-cdk/aws-codepipeline/lib/_enhanced_action.ts

```diff
@@ -0,0 +1,36 @@
+import { Aws, ResourceEnvironment } from '@aws-cdk/core';
+import { IAction } from './action';
+
+/**
+ * Helper routines to work with Actions
+ *
+ * Can't put these on Action themselves since we only have an interface
+ * and every library would need to reimplement everything (there is no
+ * `ActionBase`).
+ *
+ * So here go the members that should have gone onto the Action class
+ * but can't.
+ *
+ * It was probably my own idea but I don't want it anymore:
+ * https://github.com/aws/aws-cdk/issues/10393
+ */
+export class EnhancedAction {
+  constructor(private readonly action: IAction) {
+  }
+
+  /**
+   * The environment this action runs in
+   */
+  public get env(): ResourceEnvironment {
```

Comment:
Have to say, I don't love this class. I feel like this would be better served as a helper function, or private method in `Pipeline`.

Reply:
It is effectively a helper function, except modeled as a richer class instead of a static method. The following:

```ts
new EnhancedAction(action).env

ActionHelpers.env(action)
```

Are effectively the same thing. Except the first feels "better" to me. It's more OO. It's like [Rich Wrappers in Scala](https://docs.scala-lang.org/overviews/core/implicit-classes.html) (except explicit instead of implicit).

Reply:
I don't agree that a class whose only possible usage is `new EnhancedAction(action).env` is more OO. Classes need to have behavior, state, methods, to be considered good OO. This reminds me of the argument that instead of this:

```ts
interface Whatevs {
  value: string;
}
```

, you should do this:

```ts
class Whatevs {
  private _value: string;
  public get value(): string { return this._value; }
  public set value(value: string): void { this._value = value; }
}
```

because it's "more OO".

> ```ts
> ActionHelpers.env(action)
> ```

I don't actually think this should be a static method. This should be a private method of the `Pipeline` class, which is the entity that is actually concerned with the algorithm used to determine this (as it's what determines the support Stack algorithm).